### PR TITLE
[otbn] add standalone tests for RSA 1024

### PR DESCRIPTION
**Pull request by felixmiller**
This adds two files for testing RSA encrypt and decrypt to be in-cooperated in the CI

### CODE REVIEW

**Comment by imphil** [comment]
CI fails with the following error. @rswarbrick do you have an idea right away?

```
./otbn-as -o ../../../../build-bin/otbn/util/code-snippets/rsa_1024_dec_test.o ../../../../sw/otbn/code-snippets/rsa_1024_dec_test.S
Traceback (most recent call last):
  File "./otbn-as", line 1231, in <module>
    sys.exit(main())
  File "./otbn-as", line 1217, in main
    just_translate)
  File "./otbn-as", line 1138, in transform_inputs
    insns_file, glued_insns_dec_len, mnem_to_rve)
  File "./otbn-as", line 1110, in transform_input
    transformer.take_line(line)
  File "./otbn-as", line 1067, in take_line
    if line[pos] == '\n':
IndexError: string index out of range
Makefile:46: recipe for target '../../../../build-bin/otbn/util/code-snippets/rsa_1024_dec_test.o' failed
```

**Review by rswarbrick — APPROVED** [review]
Looks good to me, but could you add a trailing newline at the end of the files?

**Comment by rswarbrick** [comment]
Ah, this is going to need a bit of fiddling. The problem is that the code depends on `modexp.s` but the "build, assemble and link everything" CI test doesn't know how to link it.

**Comment by felixmiller** [comment]
I think the `modexp` and `modexp_65537` probably need global binding for this. Added a commit to #3813 
Would that help for the CI?

**Comment by rswarbrick** [comment]
Well, the problem is that the test currently links each object into a self-contained binary. Obviously, that's not going to work! :-) Probably the right thing to do is add a Make snippet adding dependencies. I'll try to get to it on Monday if you don't get a chance first.

**Comment by rswarbrick** [comment]
Hi @felixmiller! I've just rebased then pushed up an "add some makefile stuff" commit before your patch in this PR and amended your patch to teach make about the dependencies. Does this look sensible to you?

@imphil: Would you mind taking a look at the Makefile stuff in the 1st commit? Approving the PR myself seems a little underhand, since I wrote that bit!

**Comment by rswarbrick** [comment]
Oh, that's silly. This fails CI because one of the commits wasn't written by Felix. I'll push up the first patch as a separate PR and we can rebase this one afterwards.

**Comment by rswarbrick** [comment]
Rebase on top of #3866 (if and) when it is merged.

**Review by imphil — COMMENTED** [review]
I added some comment on the Makefiles.


Edit: Let's continue this discussion in #3866.

> @gkelly @lenary I understand that Makefiles are not a preferred solution for building SW, but I don't see any other solution coming up in the short or even medium term. So it's either Makefiles or something along the lines of a shell script.

> We can, to avoid any doubt that we're ready to use whatever build system we get once there is one for OTBN code, can we put all the Makefile snippets in hw/ip/otbn somewhere. I don't think that changes anything, however.

File: hw/ip/otbn/util/Makefile [hunk 1]
```diff
@@ -33,25 +33,16 @@ lint: $(lint-stamps)
...
-# Firstly, assemble them
-$(asm-objects): $(cs-build-dir)/%.o: \
-  $(repo-top)/sw/otbn/code-snippets/%.S otbn-as $(pylibs) | $(cs-build-dir)
-	./otbn-as -o $@ $<
-
-# Then link them
-$(linked-elfs): $(cs-build-dir)/%: \
-  $(cs-build-dir)/%.o otbn-ld $(pylibs)
-	./otbn-ld -o $@ $<
+include ../../../../sw/otbn/code-snippets/rules.mk
```
[the remark this answers is not in the document]
Comment (rswarbrick) [reply hw/ip/otbn/util/Makefile:42]: Good point. Yep!

File: sw/otbn/code-snippets/Makefile [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+# A Makefile for assembling and linking the code snippets. By default,
+# the rules generate outputs in build-bin/sw/device/otbn.
```
[the remark this answers is not in the document]
Comment (rswarbrick) [reply sw/otbn/code-snippets/Makefile:6]: Oh! I was assuming host=Ibex device=other stuff. That makes sense, shall do.

File: sw/otbn/code-snippets/Makefile [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#
+# To build into a separate directory, either set some variables and
+# include ./rules.mk in your own Makefile (see documentation there) or
+# run this Makefile, customising the following variables on the
+# command line:
+#
+#    variable   | default value
+#  -------------+----------------------------------------
+#    build-dir  | $(repo-top)/build-bin/sw/device/otbn
```
[the remark this answers is not in the document]
Comment (rswarbrick) [reply sw/otbn/code-snippets/Makefile:15]: Ah, I don't think I'd understood how that worked. I'll do that (and remove the "build-dir=..."-based configuration from here).

**Review by rswarbrick — COMMENTED** [review]
Thanks @imphil! I've updated PR #3866 with this comments. Could you have a look there to see whether the changes match what you'd expect?

**Comment by imphil** [comment]
Oh, I just realized that there's a separate PR for the Makefile changes at #3866. I'll head over to the other PR to continue to the discussion to keep this PR focused on the actual RSA code.

**Comment by rswarbrick** [comment]
Force-push rebases (on top of my change which is now merged). @felixmiller this just contains your code now, plus a couple of lines of magic Makery)

**Comment by rswarbrick** [comment]
Erk. Take 2. I'd managed to trash your signed-off-by line (sorry, Felix!). This now has you as the author and sign-off'er, which I think makes sense since you wrote all the code.

**Pull request merged**